### PR TITLE
Feature/117 update feataure layer

### LINKED ISSUE #117: Add other methods to FeatureLayer
#115 and #116 added query and attachment methods to FeatureLayer, respectively.

Add the following other methods:

- applyEdits
- clone
- createPopupTemplate
- getFeatureType
- getField
- getFieldDomain
- refresh
- save
- saveAs

### PR DESCRIPTION
**Pull request by AndersenBell**
closes #117 

This does not have Save or SaveAs, they need additional testing, so a new gissue will need to be created for it.

This branch fixes a bug with Query on the feature layer, where sending `undefined` is expected by the JS but we were creating an empty query causing unexpected results.

Making the asset copy check if directories exist before copying files

Adding support for feature layer additional functionality

### CODE REVIEW

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/CSVLayers.razor [hunk 1]
```diff
@@ -14,7 +14,7 @@
 <button disabled="@(!_mapRendered)" @onclick="AddLayer">Add new CSV Layer!</button>
 
 
-<SceneView @ref="_view" class="map-view" OnMapRendered="OnMapRendered">
+<SceneView @ref="_view" class="map-view" OnViewRendered="OnMapRendered">
```
Comment (seahro) [inline samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/CSVLayers.razor:17]: Do you think it will be good to adjust the other razor pages that use `SceneView` elements to also switch from `OnMapRendered` to `OnViewRendered` for clarity/consistency?  Like the `SceneView` in `ReactiveUtils.razor`.

**Review by seahro — APPROVED** [review]
Looks good! One question about `OnViewRendered`

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/CSVLayers.razor (hunk 1 above)
Comment (TimPurdum) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/CSVLayers.razor:17]: `OnMapRendered` is obsolete, we should update all our samples to `OnViewRendered` (that is irrespective of `MapView` vs. `SceneView`, they're both `Views`).

File: src/dymaptic.GeoBlazor.Core/Components/Layers/FeatureLayer.cs [hunk 2]
```diff
@@ -5,6 +5,7 @@
 using dymaptic.GeoBlazor.Core.Exceptions;
 using dymaptic.GeoBlazor.Core.Objects;
 using Microsoft.AspNetCore.Components;
+using Microsoft.Extensions.Options;
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/Layers/FeatureLayer.cs:8]: this doesn't seem related to other changes, could be removed?

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/Domain.cs [hunk 3]
```diff
@@ -77,19 +76,29 @@ public override async Task UnregisterChildComponent(MapComponent child)
...
     /// <summary>
     ///     The value of the code.
     /// </summary>
     [Parameter]
-    public string? Code { get; set; }
+    public T? Code { get; set; }
+}
+
+/// <summary>
+/// Inherited domains apply to domains at the feature type level. It implies that the domain for a given field at the feature
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/Widgets/Domain.cs:95]: I realize this is very nitpicky, but can we use the indented XML style I have above for all XML comments?

File: src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
+﻿using Microsoft.AspNetCore.Components;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace dymaptic.GeoBlazor.Core.Components;
+public class PortalFolder : MapComponent
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs:9]: needs class-level xml comments

Comment (TimPurdum) [reply src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs:9]: Is this class used? I can't find any references.

File: src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    /// <summary>
+    /// The date the folder was created.
+    /// </summary>
+    public DateTime CreatedDate { get; set; }
+
+
+    /// <summary>
+    ///  The portal associated with the folder.
+    /// </summary>
+    public Portal Portal { get; set; } = default!;
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs:26]: As a `MapComponent`, this would be assumed to work with nested components, so simple properties should be `[Parameter]` and `Portal` would be added via `RegisterChildComponents`.

File: src/dymaptic.GeoBlazor.Core/Scripts/dotNetBuilder.ts [hunk 6]
```diff
@@ -749,3 +752,29 @@ export function buildDotNetTimeInterval(interval: any): any | null {
...
+            dotNetDomains[domain] =  buildDotNetDomain(result.domains[domain]);
+        }
+    }
+
+    return {
+        id: result.id,
+        domains: dotNetDomains,
+        declaredClass: result.declaredClass,
+        name: result.name,
+        templates: result.templates,
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Scripts/dotNetBuilder.ts:774]: have you validated that `templates` converts OK without a builder method?

File: src/dymaptic.GeoBlazor.Core/Scripts/featureLayer.ts [hunk 7]
```diff
@@ -235,7 +255,64 @@ export default class FeatureLayerWrapper {
...
+
+        return buildDotNetFeatureType(result);
+    }
+
+    async getField(fieldName: string): Promise<DotNetField | null> {
+
+        let result = await this.layer.getField(fieldName);
+
+        if (result != undefined) {
+            let field = buildDotNetFields(Array(result));
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Scripts/featureLayer.ts:276]: this `Array(result)` could also be `[ result ]`, right? (I'm just more familiar with that pattern, this is fine).

**Review by TimPurdum — APPROVED** [review]
Looks good! A few cleanup items.

File: src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs (hunk 4 above)
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Components/PortalFolder.cs:9]: I think it is referenced when I was trying to get some things workings, but I removed it

File: src/dymaptic.GeoBlazor.Core/Scripts/featureLayer.ts (hunk 7 above)
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Scripts/featureLayer.ts:276]: There are some differences, so for consistency, I will use [ ]

File: src/dymaptic.GeoBlazor.Core/Scripts/dotNetBuilder.ts (hunk 6 above)
Comment (AndersenBell) [reply src/dymaptic.GeoBlazor.Core/Scripts/dotNetBuilder.ts:774]: yes
